executor/insert_common.go

```diff
@@ -690,28 +689,15 @@ func (e *InsertValues) fillRow(ctx context.Context, row []types.Datum, hasValue
 			}
 		}
 	}
-	tbl := e.Table.Meta()
+
 	// Handle exchange partition
-	if tbl.ExchangePartitionInfo != nil {
-		is := e.Ctx().GetDomainInfoSchema().(infoschema.InfoSchema)
-		pt, tableFound := is.TableByID(tbl.ExchangePartitionInfo.ExchangePartitionID)
-		if !tableFound {
-			return nil, errors.Errorf("exchange partition process table by id failed")
-		}
-		p, ok := pt.(table.PartitionedTable)
-		if !ok {
-			return nil, errors.Errorf("exchange partition process assert table partition failed")
-		}
-		err := p.CheckForExchangePartition(
-			e.Ctx(),
-			pt.Meta().Partition,
-			row,
-			tbl.ExchangePartitionInfo.ExchangePartitionDefID,
-		)
-		if err != nil {
+	tbl := e.Table.Meta()
+	if tbl.ExchangePartitionInfo != nil && tbl.GetPartitionInfo() == nil {
+		if err := checkRowForExchangePartition(e.Ctx(), row, tbl); err != nil {
 			return nil, err
 		}
 	}
+
```

Comment:
Does this check need to be here, or could it be moved to `AddRecord` [just before checking the constraints](https://github.com/pingcap/tidb/blob/1159bc8d88c4b35a957087813d1087fbcd519725/table/tables/tables.go#L970)?

Reply:
I think it it not a very bad way to put it here because `AddRecord ` is  a function.of physical table.  In ExchangePartition case, both non-partition table and physical table of partition table will go here right? There seems no good method to distingrish it.

Reply:
OK, let's keep it there. I do think the TableCommon.AddRecord could be used for both partitioned and non-partitioned tables, but that can be a follow up issue/PR, which is not necessary.